Source/JavaScriptCore/dfg/DFGPutStackSinkingPhase.cpp

```diff
@@ -504,8 +504,11 @@ class PutStackSinkingPhase : public Phase {
                         Node* incoming = mapping.operand(operand);
                         DFG_ASSERT(m_graph, node, incoming);
                     
+                        // If we are sinking a PutStack to before an ExitOK, it
+                        // should be ExitInvalid like the other preceding nodes.
+                        auto origin = node->op() == ExitOK ? node->origin.withInvalidExit() : node->origin;
```

Comment:
What is `ExitOK` node? How do I understand a `ExitOK` node with node origin with invalid exit?

Reply:
`ExitOK` is an indicator we add to the graph after a run of instructions where OSR exit is not valid. For example, if a basic block contains `Phi` nodes, we don't want to exit until those `Phi`s and their associated `MovHint` (and now `PutStack`) instructions have executed, setting up the OSR exit buffer. After these nodes, we generate an `ExitOK` to indicate that OSR exit is allowed.

In this case, this is relevant because in the PutStack sinking phase, when we add a `PutStack`, the existing implementation just uses whatever the origin of the instruction we're inserting before was. This works in most cases, but specifically when we have an `ExitOK` (where exit is valid), we would add a `PutStack` with `ExitValid` _before_ the `ExitOK`. This isn't legal (although it's unlikely to cause real issues at runtime) and we would fail validation. So, this patch makes it so if we are inserting directly before an `ExitOK`, that the new `PutStack` has `ExitInvalid`.

Reply:
To clarify, the ExitOK node is only needed in places where we go from OSR exiting being invalid to valid within the same `BytecodeIndex`. If we go from exit invalid to valid between bytecodes/checkpoints that's OK.